README.md
# Glowe

Julia interface to [GloVe](https://nlp.stanford.edu/projects/glove/).

[![License](http://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat)](LICENSE.md)
[![Build Status](https://travis-ci.org/zgornel/Glowe.jl.svg?branch=master)](https://travis-ci.org/zgornel/Glowe.jl)
[![Coverage Status](https://coveralls.io/repos/github/zgornel/Glowe.jl/badge.svg?branch=master)](https://coveralls.io/github/zgornel/Glowe.jl?branch=master)

This package provides functionality for generating and working with GloVe word embeddings. The training is done using the original C code from the [GloVe github repository](https://github.com/stanfordnlp/GloVe).

Note that there is also a package called [Glove.jl](https://github.com/domluna/Glove.jl) that provides a pure Julia implementation of the algorithm.

* [Release Notes](https://github.com/zgornel/Glowe.jl/blob/master/NEWS.md)


## Installation

```julia
Pkg.clone("https://github.com/zgornel/Glowe.jl")
```
for the latest `master` or
```julia
Pkg.add("Glowe")
```
for the stable versions.


## Documentation

Most of the documentation is provided in Julia's native docsystem.


## Examples

Following Word2Vec.jl's example, considering the corpus from http://mattmahoney.net/dc/text8.zip extracted as text file ``text8`` in the current working directory, the GloVe model can be obtained with:

```julia
julia> # Training (may take a while)
       vocab_count("text8", "vocab.txt", min_count=5, verbose=1);
       cooccur("text8", "vocab.txt", "cooccurrence.bin", memory=8.0, verbose=1);
       shuffle("cooccurrence.bin", "cooccurrence.shuf.bin", memory=8.0, verbose=1);
       glove("cooccurrence.shuf.bin", "vocab.txt", "text8-vec", threads=8,
             x_max=10.0, iter=15, vector_size=300, binary=0, write_header=1,
             verbose=1);
# BUILDING VOCABULARY
# Truncating vocabulary at min count 5.
# Using vocabulary of size 71290.
#
# COUNTING COOCCURRENCES
# window size: 15
# context: symmetric
# Merging cooccurrence files: processed 60666468 lines.
#
# SHUFFLING COOCCURRENCES
# array size: 510027366
# Merging temp files: processed 60666468 lines.
#
# TRAINING MODEL
# Read 60666468 lines.
# vector size: 300
# vocab size: 71290
# x_max: 10.000000
# alpha: 0.750000
# 12/11/18 - 12:58.58AM, iter: 001, cost: 0.070201
# 12/11/18 - 01:00.33AM, iter: 002, cost: 0.052521
# ...
```

The model can be imported with
```julia
model = wordvectors("text8-vec.txt", Float32, header=true, kind=:text)
# WordVectors 71290 words, 300-element Float32 vectors
```

The vector representation of a word can be obtained using ``get_vector``.
```julia
julia> get_vector(model, "book")'
# 300-element Array{Float32,1}:
#   0.006189716
#   0.04822071
#   0.017121462
#   ...
```

The cosine similarity of ``book``, for example, can be computed using ``cosine_similar_words``.
```julia
julia> cosine_similar_words(model, "book")
# 10-element Array{String,1}:
#  "book"
#  "books"
#  "published"
#  "domesday"
#  "novel"
#  "comic"
#  "written"
#  "bible"
#  "urantia"
#  "work"
```

Word vectors have many interesting properties. For example,
``vector("king") - vector("man") + vector("woman")`` is close to ``vector("queen")``.

```julia
julia> analogy_words(model, ["king", "woman"], ["man"])
# 5-element Array{String,1}:
#  "queen"
#  "daughter"
#  "children"
#  "wife"
#  "son"
```


## License

This code has an MIT license and therefore it is free.
GloVe is released under an Apache License v2.0.


## References

[1] [GloVe: Global Vectors for Word Representation](https://nlp.stanford.edu/projects/glove/)

[2] [Glove.jl - native Julia implementation](https://github.com/domluna/Glove.jl)


## Acknowledgements

The design of the package relies on design concepts from [the word2vec Julia interface, Word2Vec.jl](https://github.com/zgornel/Word2Vec.jl).


## Reporting Bugs

Please [file an issue](https://github.com/zgornel/Glowe.jl/issues/new) to report a bug or request a feature.


NEWS.md
## Glowe Release Notes

v0.1.0
------
Initial release


test/model.jl
# test various model functions
bigdata = joinpath(dirname(@__FILE__), "..", "data", "big.txt")

TMPDIR = "./tmp"
mkpath(TMPDIR)

CORPUS = bigdata
VOCAB_FILE = joinpath(TMPDIR, "vocab.txt")
COOCCURRENCE_FILE = joinpath(TMPDIR, "cooccurrence.bin")
COOCCURRENCE_SHUF_FILE = joinpath(TMPDIR, "cooccurrence.shuf.bin")
SAVE_FILE = joinpath(TMPDIR, "vectors")
VERBOSE = 2
MEMORY = 4.0
VOCAB_MIN_COUNT = 5
VECTOR_SIZE = 50
MAX_ITER = 3
WINDOW_SIZE = 5
BINARY= 0
NUM_THREADS = 1
X_MAX = 10.0
HEADER = 1

# Create and load model
vocab_count(CORPUS, VOCAB_FILE, min_count=VOCAB_MIN_COUNT, verbose=VERBOSE)
cooccur(CORPUS, VOCAB_FILE, COOCCURRENCE_FILE,
        memory=MEMORY, verbose=VERBOSE, window_size=WINDOW_SIZE)
shuffle(COOCCURRENCE_FILE, COOCCURRENCE_SHUF_FILE,
memory=MEMORY, verbose=VERBOSE)
glove(COOCCURRENCE_SHUF_FILE, VOCAB_FILE, SAVE_FILE,
      threads=NUM_THREADS, x_max=X_MAX, iter=MAX_ITER,
      vector_size=VECTOR_SIZE, binary=BINARY,
      write_header=HEADER, verbose=VERBOSE)
model_file = joinpath(SAVE_FILE) * ifelse(BINARY==1, ".bin", ".txt")
model = wordvectors(model_file, Float32, header=true, kind= :text)

# word vectors
println(model)

len_vecs, num_words = size(model)
wordvecs = model.vectors
@test size(wordvecs) == (len_vecs, num_words)

words = vocabulary(model)
word1 = words[rand(1:end)]
word2 = words[rand(1:end)]
word3 = words[rand(1:end)]

@test in_vocabulary(model, word1)

n = rand(1:100)

indxs, mes = cosine(model, word1, n)
@test words[indxs] == cosine_similar_words(model, word1, n)
w4_indx = indxs[rand(1:end)]
loc = findall(in(w4_indx), indxs)
word4 = words[w4_indx]
@test index(model, word4) == w4_indx

s = similarity(model, word1, word4)
@test mes[loc[1]] ≈ s

inx, mes = analogy(model, [word1, word2], [word3], n)
@test words[inx] == analogy_words(model, [word1, word2], [word3], n)

rm(TMPDIR, recursive=true, force=true)

println("model passed test...")


test/train.jl
# test various training parameters
bigdata = joinpath(dirname(@__FILE__), "..", "data", "big.txt")

TMPDIR = "./tmp"
mkpath(TMPDIR)

CORPUS = bigdata
VOCAB_FILE = joinpath(TMPDIR, "vocab.txt")
COOCCURRENCE_FILE = joinpath(TMPDIR, "cooccurrence.bin")
COOCCURRENCE_SHUF_FILE = joinpath(TMPDIR, "cooccurrence.shuf.bin")
SAVE_FILE = joinpath(TMPDIR, "vectors")
VERBOSE = 0
MEMORY = 4.0
VOCAB_MIN_COUNT = 5
VECTOR_SIZE = 13
MAX_ITER = 2
WINDOW_SIZE = 3
BINARY_OPTS = [0,1]
NUM_THREADS = 1
X_MAX = 10.0
HEADER_OPTS = [0,1]

vocab_count(CORPUS, VOCAB_FILE, min_count=VOCAB_MIN_COUNT, verbose=VERBOSE)
cooccur(CORPUS, VOCAB_FILE, COOCCURRENCE_FILE,
        memory=MEMORY, verbose=VERBOSE, window_size=WINDOW_SIZE)
shuffle(COOCCURRENCE_FILE, COOCCURRENCE_SHUF_FILE,
        memory=MEMORY, verbose=VERBOSE)
for BINARY in BINARY_OPTS
    for HEADER in HEADER_OPTS
        glove(COOCCURRENCE_SHUF_FILE, VOCAB_FILE, SAVE_FILE,
              threads=NUM_THREADS, x_max=X_MAX, iter=MAX_ITER,
              vector_size=VECTOR_SIZE, binary=BINARY,
              write_header=HEADER, verbose=VERBOSE)
        model_file = joinpath(SAVE_FILE) * ifelse(BINARY==1, ".bin", ".txt")
        TYPE = Float32
        if BINARY == 0
            model = wordvectors(model_file,
                                TYPE,
                                header=(HEADER==1),
                                kind=:text)
        else
            model = wordvectors(model_file,
                                TYPE,
                                kind=:binary,
                                vocabulary=VOCAB_FILE)
        end
        len_vecs, num_words = size(model)
        @test model.vectors isa Matrix{TYPE}
        @test len_vecs == VECTOR_SIZE
    end
end

rm(TMPDIR, recursive=true, force=true)

println("training passed test...")



test/runtests.jl
using Test
using Glowe

@testset "Model training" begin
    include("train.jl")
end


src/Glowe.jl
#                                                  ....
#                                 ......''''''....;cll:;,..
#                          .';cllc:::cccc:c::::cccldxdlll:,.
#                         .ldxxxxkdlcccc:ccccccccclodoccllc.
#                       .':oddkxxlxkdollccccccc::::cl:';c:'.
#                       ,ccdxkkxd;dOxddoclllcllllc:::c,....
#                      .;llcclc:;:xxxddooxkkddddol:::::;;;'.
#                    ..,;cloollooddddddddxxkO0Odollcc:;;,'.
#                 .,;lddl:cloooddddddddxkOOOOOkxdolc:;,'..
#          .;:,. .:loxOOdlccoodxxxxxxxdddxxxxddolc;,'...
#        .l0K0ko:coodOK0xol;;,;::ccc:;;;;;;;;,''........    ..:dxl,
#       .l0XX0xolodddkOkdolc:::,'.......................  ..,:oO00Oc
#        :kOkxoloxxkkkdc;;;;;;,,''.........................':lodxkx:.
#        .odooolxxxkkkd,;clcccc;,'.......................',:loool:'
#      .;:ododxkxdkkkkx;;olllccc:;,,,,,,,,'''''''...;c:;;;:clc;.
#      ,ldxxxkkkddxkxxxoclclclcc:::::::cc:::::;,'..cxkoc:;;;.
#      'ldxxxxkdlddxkxkOkxxo::c:::::;;;;;::cl:,'.'ckOOdc;'.....
#      .lddddddxxdkkOxxO0kdxOkl,'.......,:lodooloodkkxo:;:ccllcl:;.
#      .cdoc:oxkOOkkOkxk0kxxdxkdo:.....,dkkxxxdoxxxxxdc;;lddxdddol;.
#       'ld:..:lxdxOOkxkkxxl'okxdc.....'lxkdlcdkOOxk0Ol,.:oddddxdcc.
#        ,:.  .;kOOkkOxxkkx:..,;,.......',,,'ckO0kxO0k:...',oddxocl;..  ........
#            .cxkxolc;:xkkd:................ckkkkxxxdl,....'lddolco:'...........
#            .lololc..ckkxxl'..............cxxddxdddl:'....'ldol:lo:..........
#             .,,;'. .oxxxxd:............,lxdoc;,coo:'.....'ldc;loc,........
#                     .,:::'..   ..  . .'oddol;':ooc;.....,cc:.'ll:'..    ..
#                                       .,;:;,..;c:;'. . 'cl;. .....
#                                                 .      ...
# Glowe.jl - Julia interface to GloVe
#            Written in Julia at 0x0α Research
#			 by Corneliu Cofaru, 2018
#
# Original authors: Jeffrey Pennington, Richard Socher
#					and Christopher D. Manning. 2014

module Glowe

    using LinearAlgebra
    using Statistics
    import Base: show, size

    export vocab_count, cooccur, shuffle, glove,
           WordVectors, wordvectors,
		   index, size,
		   get_vector,
           vocabulary, in_vocabulary,
           cosine, cosine_similar_words,
           analogy, analogy_words,
		   similarity

    include("interface.jl")
    include("wordvectors.jl")

end # module


src/wordvectors.jl
mutable struct WordVectors{S<:AbstractString, T<:Real, H<:Integer}
    vocab::Vector{S} # vocabulary
    vectors::Array{T, 2} # the vectors computed from GloVe
    vocab_hash::Dict{S, H}
end

function WordVectors(vocab::AbstractArray{S,1},
                     vectors::AbstractArray{T,2}) where {S<: AbstractString, T<:Real}
    length(vocab) == size(vectors, 2) ||
        throw(DimensionMismatch("Dimension of vocab and vectors are inconsistent."))
    vocab_hash = Dict{S, Int}()
    for (i, word) in enumerate(vocab)
        vocab_hash[word] = i
    end
    WordVectors(vocab, vectors, vocab_hash)
end


function Base.show(io::IO, wv::WordVectors{S,T}) where {S,T}
    len_vecs, num_words = size(wv.vectors)
    print(io, "WordVectors $(num_words) words, $(len_vecs)-element $(T) vectors")
end


"""
    vocabulary(wv)

Return the vocabulary as a vector of words of the WordVectors `wv`.
"""
vocabulary(wv::WordVectors) = wv.vocab


"""
    in_vocabulary(wv, word)

Return `true` if `word` is part of the vocabulary of the WordVector `wv` and
`false` otherwise.
"""
in_vocabulary(wv::WordVectors, word::AbstractString) = word in wv.vocab


"""
    size(wv)

Return the word vector length and the number of words as a tuple.
"""
size(wv::WordVectors) = size(wv.vectors)


"""
    index(wv, word)

Return the index of `word` from the WordVectors `wv`.
"""
index(wv::WordVectors, word) = wv.vocab_hash[word]


"""
    get_vector(wv, word [; oov=false, oov_key="<unk>")

Returns the vector representation of `word` from the WordVectors `wv`.
If `oov` is `true`, the vector corresponding to the key `oov_key` is
returned for out-of-vocabulary words.
"""
get_vector(wv::WordVectors,
           word;
           oov::Bool=false,
           oov_key::String="<unk>") = begin
    if !oov
        idx = wv.vocab_hash[word];
        return wv.vectors[:, idx]
    else
        idx = get(wv.vocab_hash, word, wv.vocab_hash[oov_key])
        return wv.vectors[:, idx]
    end
end


"""
    cosine(wv, word, n=10)

Return the position of `n` (by default `n = 10`) neighbors of `word` and their
cosine similarities.
"""
function cosine(wv::WordVectors, word, n=10)
    metrics = wv.vectors'*get_vector(wv, word)
    topn_positions = sortperm(metrics[:], rev = true)[1:n]
    topn_metrics = metrics[topn_positions]
    return topn_positions, topn_metrics
end


"""
    similarity(wv, word1, word2)

Return the cosine similarity value between two words `word1` and `word2`.
"""
function similarity(wv::WordVectors, word1, word2)
    return get_vector(wv, word1)'*get_vector(wv, word2)
end


"""
    cosine_similar_words(wv, word, n=10)

Return the top `n` (by default `n = 10`) most similar words to `word`
from the WordVectors `wv`.
"""
function cosine_similar_words(wv::WordVectors, word, n=10)
    indx, metr = cosine(wv, word, n)
    return vocabulary(wv)[indx]
end


"""
    analogy(wv, pos, neg, n=5)

Compute the analogy similarity between two lists of words. The positions
and the similarity values of the top `n` similar words will be returned.
For example,
`king - man + woman = queen` will be
`pos=[\"king\", \"woman\"], neg=[\"man\"]`.
"""
function analogy(wv::WordVectors{S,T,H}, pos::AbstractArray, neg::AbstractArray, n= 5
                ) where {S<:AbstractString, T<:Real, H<:Integer}
    m, n_vocab = size(wv)
    n_pos = length(pos)
    n_neg = length(neg)
    anal_vecs = Matrix{T}(undef, m, n_pos + n_neg)

    for (i, word) in enumerate(pos)
        anal_vecs[:,i] = get_vector(wv, word)
    end
    for (i, word) in enumerate(neg)
        anal_vecs[:,i+n_pos] = -get_vector(wv, word)
    end
    mean_vec = mean(anal_vecs, dims=2)
    metrics = wv.vectors'*mean_vec
    top_positions = sortperm(metrics[:], rev = true)[1:n+n_pos+n_neg]
    for word in [pos;neg]
        idx = index(wv, word)
        loc = findfirst(x->x==idx, top_positions)
        if loc != nothing
            splice!(top_positions, loc)
        end
    end
    topn_positions = top_positions[1:n]
    topn_metrics = metrics[topn_positions]
    return topn_positions, topn_metrics
end


"""
    analogy_words(wv, pos, neg, n=5)

Return the top `n` words computed by analogy similarity between
positive words `pos` and negaive words `neg`. from the WordVectors `wv`.
"""
function analogy_words(wv::WordVectors, pos, neg, n=5)
    indx, metr = analogy(wv, pos, neg, n)
    return vocabulary(wv)[indx]
end


"""
    wordvectors(filename [,type=Float64][; kind=:text, header=false, normalize=true, vocabulary=nothing])

Generate a WordVectors type object from a file.

# Arguments
  * `filename::AbstractString` the embeddings file name
  * `type::Type` type of the embedding vector elements; default `Float64`

# Keyword arguments
  * `kind::Symbol` specifies whether the embeddings file is textual (`:text`)
or binary (`:binary`); default `:text`
  * `header::Union{Nothing, Bool}` in text embeddings files specifies
  whether the file contains a header i.e. number of lines, columns or not.
  If the header is `nothing`, the loader will attempt to autodetect the
  presence of a header; default `nothing`
  * `normalize:Bool` specifies whether to normalize the embedding vectors
i.e. return unit vectors; default true
  * `vocabulary::Union{Nothing, AbstractString}` path to the vocabulary
file generated with `vocab_count` (needed for binary embeddings);
default `nothing`
  * `load_bias::Bool` specifies whether to load the bias term or not
if using binary embedding files; default `false`
"""
function wordvectors(filename::AbstractString,
                     ::Type{T};
                     kind::Symbol=:text,
                     header::Union{Nothing, Bool}=nothing,
                     normalize::Bool=true,
                     vocabulary::Union{Nothing, AbstractString}=nothing,
                     load_bias::Bool=false
                    ) where T <: Real
    if kind == :binary
        vocabulary == nothing &&
            @error "A GloVe-generated vocabulary file path must be specified."
        return _from_binary(T, filename, vocabulary, load_bias, normalize)
    elseif kind == :text
        header == nothing && (header = autodetect_header(filename))
        if header
            return _from_text_header(T, filename, normalize)
        else
            return _from_text(T, filename, normalize)
        end
    else
        throw(ArgumentError("Unknown embedding file kind $(kind)"))
    end
end


"""
    autodetect_header(filename)

Function that attempts at autodetecting the presence of a header in a
GloVe embeddings file in a text format. If the function fails to detect
a header, a `false` value is returned.

Note: The function explicitly expects a text format for the embeddings.
The behaviour is undetermined for binary formats and no attempt at
detecting the file format is done.
"""
function autodetect_header(filename::AbstractString)
    open(filename) do f
        tokens1 = split(strip(readline(f)), ' ')
        tokens2 = split(strip(readline(f)), ' ')
        try
            # Attempt to parse the first line as header,
            # the second as an embedding
            @assert length(tokens1) == 2
            vocab_size = parse(Int, tokens1[1])
            vector_size = parse(Int, tokens1[2])
            vector = map(x-> parse(Float64, x), tokens2[2:end])
            @assert (vector_size == 2*(length(vector)+1) ||  # `model` is 0
                     vector_size == length(vector))          # `model` is 1 or 2
            return true
        catch
            return false
        end
    end
end


wordvectors(filename::AbstractString;
            kind::Symbol=:text,
            header::Bool=false,
            normalize::Bool=true,
            vocabulary::Union{Nothing, AbstractString}=nothing) =
    wordvectors(filename, Float64, kind=kind, header=header,
                normalize=normalize, vocabulary=vocabulary)


# Generate a WordVectors object from binary file
function _from_binary(::Type{T},
                      filename::AbstractString,
                      vocabulary::AbstractString,
                      load_bias::Bool=false,
                      normalize::Bool=true) where T<:Real
    vocab = Vector{String}(undef, 0)
    open(vocabulary) do f
        for line in eachline(f)
            word, count = split(line, ' ')
            push!(vocab, word)
        end
    end
    vocab_size = length(vocab)
    open(filename) do f
        nb = 0  # byte count
        while !eof(f)
            read(f,1)
            nb+= 1
        end
        vector_size = Int(nb / (2 * 8 * vocab_size))
        vector_mask = 1 : (vector_size - ifelse(load_bias, 0, 1))
        vectors = Matrix{T}(undef, length(vector_mask), vocab_size)
        binary_length = sizeof(Float64) * vector_size
        seekstart(f)
        for i in 1:vocab_size
            vector = reinterpret(Float64, read(f, binary_length))[vector_mask]
            if normalize
                vector = vector ./ norm(vector)  # unit vector
            end
            vectors[:,i] = T.(vector)
        end
        return WordVectors(vocab, vectors)
    end
end


# Generate a WordVectors object from text file
function _from_text_header(::Type{T},
                           filename::AbstractString,
                           normalize::Bool=true) where T<:Real
    open(filename) do f
        header = strip(readline(f))
        vocab_size, vector_size = map(x -> parse(Int, x), split(header, ' '))
        vocab = Vector{String}(undef, vocab_size)
        vectors = Matrix{T}(undef, vector_size, vocab_size)
        i = 1
        for line in eachline(f)
            if i > 1
                line = strip(line)
                parts = split(line, ' ')
                word = parts[1]
                vector = map(x-> parse(T, x), parts[2:end])
                if normalize
                    vector = vector ./ norm(vector)  # unit vector
                end
                vocab[i-1] = word
                vectors[:, i-1] = vector
            end
            i+= 1
        end
    return WordVectors(vocab, vectors)
    end
end


# Generate a WordVectors object from text file
function _from_text(::Type{T},
                    filename::AbstractString,
                    normalize::Bool=true) where T<:Real
    open(filename) do f
        vocab = Vector{String}()
        vectors = Vector{Vector{T}}()
        for (i, line) in enumerate(readlines(f))
            line = strip(line)
            parts = split(line, ' ')
            word = parts[1]
            vector = map(x-> parse(T, x), parts[2:end])
            if normalize
                vector = vector ./ norm(vector)  # unit vector
            end
            push!(vocab, word)
            push!(vectors, vector)
        end
        return WordVectors(vocab, reduce(hcat, vectors))
    end
end


src/interface.jl
"""
    vocab_count(corpus, vocab; max_vocab=100_000, min_count=10, verbose=2)

Extracts unigram counts.

# Arguments
  * `corpus::AbstractString` input corpus file path
  * `vocab::AbstractString` output vocabulary file path

# Keyword arguments
  * `verbose::Int` set verbosity: 0, 1 or 2 (default)
  * `max_vocab::Int` upper bound on vocabulary size, i.e. keep the <int> most frequent words. The minimum frequency words are randomly sampled so as to obtain an even distribution over the alphabet.
  * `min_count::Int` lower limit such that words which occur fewer than <int> times are discarded.

# Examples
```
julia> vocab_count("corpus.txt", "vocab.txt", verbose=2, max_vocab=100_000, min_count=10)
```
"""
function vocab_count(corpus::AbstractString,
                     vocab::AbstractString;
                     max_vocab::Int=100_000,
                     min_count::Int=10,
                     verbose::Int=2)
    command = joinpath(dirname(@__FILE__), "..",
                       "deps", "src", "GloVe-c", "build", "./vocab_count")
    parameters = String[]
    args = ["-max-vocab", "-min-count", "-verbose"]
    values = [max_vocab, min_count, verbose]

    for (arg, value) in zip(args, values)
        push!(parameters, arg)
        push!(parameters, string(value))
    end

    run(pipeline(pipeline(corpus, `$command $parameters`), vocab))
    return nothing
end


"""
    cooccur(corpus, vocab, cooccurrences; verbose=2, symmetric=0, window_size=10, memory=4.0, max_product=nothing, overflow_length=nothing, overflow_file="overflow", distance_weighting=1)

Calculates word-word cooccurrence statistics.

# Arguments
  * `corpus::AbstractString` input corpus file path
  * `vocab::AbstractString` input vocabulary file path (the vocabulary contains truncated unigram counts, produced by `vocab_count`)
  * `cooccurrences::AbstractString` output cooccurrences file path

# Keyword arguments
  * `verbose::Int` set verbosity: 0, 1, 2 (default) or 3
  * `symmetric::Int` if <int> = 0, only use left context; if <int> = 1 (default), use left and right
  * `window_size::Int` number of context words to the left (and to the right, if symmetric = 1); default 15
  * `memory::Float64` soft limit for memory consumption, in GB -- based on simple heuristic, so not extremely accurate; default 4.0
  * `max_product::Union{Nothing, Int}` limit the size of dense cooccurrence array by specifying the max product <int> of the frequency counts of the two cooccurring words. This value overrides that which is automatically produced by `memory`. Typically only needs adjustment for use with very large corpora
  * `overflow_length::Union{Nothing, Int}` limit to length <int> the sparse overflow array, which buffers cooccurrence data that does not fit in the dense array, before writing to disk. This value overrides that which is automatically produced by `memory`. Typically only needs adjustment for use with very large corpora
  * `overflow_file::String` filename, excluding extension, for temporary files; default "overflow"
  * `distance_weighting::Int` if <int> = 0, do not weight cooccurrence count by distance between words; if <int> = 1 (default), weight the cooccurrence count by inverse of distance between words"

# Examples
```
# It is assumed that vocab.txt exists and has been created by `vocab_count`
julia> cooccur("corpus.txt", "vocab.txt", "cooccurrences.bin", verbose=2, symmetric=0, window_size=10, memory=8.0, overflow_file="tempoverflow")
```
"""
function cooccur(corpus::AbstractString,
                 vocab::AbstractString,
                 cooccurrences::AbstractString;
                 verbose::Int=2,
                 symmetric::Int=1,
                 window_size::Int=15,
                 memory::Float64=4.0,
                 max_product::Union{Nothing, Int}=nothing,
                 overflow_length::Union{Nothing, Int}=nothing,
                 overflow_file::String="overflow",
                 distance_weighting::Int=1)
    command = joinpath(dirname(@__FILE__), "..",
                       "deps", "src", "GloVe-c", "build", "./cooccur")
    parameters = String[]
    args = ["-verbose", "-symmetric", "-window-size", "-memory",
            "-max-product", "-overflow-length", "-overflow-file",
            "-distance-weighting"]
    values = [verbose, symmetric, window_size, memory,
              max_product, overflow_length, overflow_file,
              distance_weighting]

    for (arg, value) in zip(args, values)
        if value != nothing
            push!(parameters, arg)
            push!(parameters, string(value))
        end
    end
    # Push '-vocab-file' separately
    push!(parameters, "-vocab-file")
    push!(parameters, vocab)

    run(pipeline(pipeline(corpus, `$command $parameters`), cooccurrences))
    return nothing
end


"""
    shuffle(cooccurences, shuffled; verbose=2, memory=4.0, array_size=nothing, temp_file="temp_shuffle")

Shuffles entries of word-word cooccurrence files.

# Arguments
  * `cooccurences::AbstractString` input cooccurrences file path
  * `shuffled::AbstractString` output shuffled cooccurences file path

# Keyword arguments
  * `verbose::Int` set verbosity: 0, 1, or 2 (default)
  * `memory::Float64` soft limit for memory consumption, in GB; default 4.0
  * `array_size::Union{Nothing, Int}` limit to length <int> the buffer which stores chunks of data to shuffle before writing to disk. This value overrides that which is automatically produced by `memory`
  * `temp_file`::String filename, excluding extension, for temporary files; default "temp_shuffle"

# Examples
```
# It is assumed that cooccurences.bin exists and has been created by `cooccur`
julia> shuffle("cooccurences.bin", "shuffled.bin", verbose=2, memory=8.0)
```
"""
function shuffle(cooccurrences::AbstractString,
                 shuffled::AbstractString;
                 verbose::Int=2,
                 memory::Float64=4.0,
                 array_size::Union{Nothing, Int}=nothing,
                 temp_file::String="temp_shuffle")
    command = joinpath(dirname(@__FILE__), "..",
                       "deps", "src", "GloVe-c", "build", "./shuffle")
    parameters = String[]
    args = ["-verbose", "-memory", "-array-size", "-temp-file"]
    values = [verbose, memory, array_size, temp_file]

    for (arg, value) in zip(args, values)
        if value != nothing
            push!(parameters, arg)
            push!(parameters, string(value))
        end
    end

    run(pipeline(pipeline(cooccurrences, `$command $parameters`), shuffled))
    return nothing
end


"""
    glove(shuffled, vocab, vectors="vectors"; verbose=2, write_header=0, vector_size=50, threads=8, iter=25, eta=0.05, alpha=0.75, x_max=100.0, binary=0, model=2, gradsq_file="gradsq", save_gradsq=0, checkpoint_every=0)

GloVe: Global Vectors for Word Representation, v0.2

# Arguments
  * `shuffled::AbstractString` binary input file of shuffled cooccurrence data (produced by `cooccur` and `shuffle`)
  * `vocab::AbstractString` file containing vocabulary (truncated unigram counts, produced by `vocab_count`)
  * `vectors::AbstractString` filename, excluding extension, for word vector output; default "vectors"

# Keyword arguments
  * `verbose::Int` set verbosity: 0, 1, or 2 (default)
  * `write_header::Int` if 1, write vocab_size/vector_size as first line. Do nothing if 0 (default)
  * `vector_size::Int` dimension of word vector representations (excluding bias term); default 50
  * `threads::Int` number of threads; default 8
  * `iter::Int` number of training iterations; default 25
  * `eta::Float64` initial learning rate; default 0.05
  * `alpha::Float64` parameter in exponent of weighting function; default 0.75
  * `x_max::Float64` parameter specifying cutoff in weighting function; default 100.0
  * `binary::Int` save output in binary format (0: text, 1: binary, 2: both); default 0
  * `model::Int` model for word vector output (for text output only); default 2
        0: output all data, for both word and context word vectors, including bias terms
        1: output word vectors, excluding bias terms
        2: output word vectors + context word vectors, excluding bias terms
  * `gradsq_file::String` filename, excluding extension, for squared gradient output; default "gradsq"
  * `save_gradsq::Int` save accumulated squared gradients; default 0 (off); ignored if `gradsq_file`
  * `checkpoint_every::Int` checkpoint a  model every <int> iterations; default 0 (off)

# Examples
```
# It is assumed that:
#  - shuffled.bin exists and has been created by `shuffled`
#  - vocab.txt exists and has been created by `vocab_count`
julia> glove("shuffled.bin", "vocab.txt", "vectors", gradsq_file="gradsq", verbose=2, vector_size=100, threads=16, alpha=0.75, x_max=100.0, eta=0.05, binary=2, model=2)
```
"""

#printf("./glove -input-file cooccurrence.shuf.bin -vocab-file vocab.txt -save-file vectors -gradsq-file gradsq -verbose 2 -vector-size 100 -threads 16 -alpha 0.75 -x-max 100.0 -eta 0.05 -binary 2 -model 2\n\n");
function glove(shuffled::AbstractString,
               vocab::AbstractString,
               vectors::AbstractString="vectors";
               verbose::Int=2,
               write_header::Int=0,
               vector_size::Int=50,
               threads::Int=8,
               iter::Int=25,
               eta::Float64=0.05,
               alpha::Float64=0.75,
               x_max::Float64=100.0,
               binary::Int=0,
               model::Int=2,
               gradsq_file::Union{Nothing, String}=nothing,
               save_gradsq::Int=0,
               checkpoint_every::Int=0)
    command = joinpath(dirname(@__FILE__), "..",
                       "deps", "src", "GloVe-c", "build", "./glove")
    parameters = String[]
    args = ["-verbose", "-write-header", "-vector-size", "-threads", "-iter",
            "-eta", "-alpha", "-x-max", "-binary", "-model", "-save-gradsq",
            "-checkpoint-every"]
    values = [verbose, write_header, vector_size, threads, iter,
              eta, alpha, x_max, binary, model, save_gradsq,
              checkpoint_every]

    for (arg, value) in zip(args, values)
        push!(parameters, arg)
        push!(parameters, string(value))
    end

    # Handle gradsq_file and save_gradsq
    if gradsq_file != nothing
        push!(parameters, "-gradsq-file")
        push!(parameters, string(gradsq_file))
    end
    if gradsq_file == nothing && save_gradsq != 0
        push!(parameters, "-save-gradsq")
        push!(parameters, string(save_gradsq))
        push!(parameters, "-gradsq-file")
        push!(parameters, "gradsq")
    end
    # Handle input and output arguments
    push!(parameters, "-input-file")
    push!(parameters, shuffled)
    push!(parameters, "-vocab-file")
    push!(parameters, vocab)
    push!(parameters, "-save-file")
    push!(parameters, vectors)

    run(`$command $parameters`)
    return nothing
end
